index.bs

```diff
@@ -679,11 +680,91 @@ are traditionally added to {{WindowEventHandlers}} rather than {{Window}}.</p>
+    for guarding against unplanned recursion. 
+    
+    In this example, an object of type <dfn noexport interface>MyObject</dfn> has
+    a method <dfn method for="MyObject">doComplexOpWithEvents()</dfn>. {{MyObject}}
+    also has an internal state attribute
+    {{MyObject}}.[[<dfn noexport attribute for="MyObject">started</dfn>]] whose initial
+    value is `false`.
+
+    The {{doComplexOpWithEvents()}} method MUST act as follows:
+    
+    1. If {{MyObject}}.[[{{started}}]] is not `false` then throw an {{InvalidStateError}}
+        {{DOMException}} and terminate this algorithm.
+        
+        Note: {{doComplexOpWithEvents()}} is not allowed to be run again
+        (e.g., from within a `"currentlyinprogress"` event handler) before it finishes
+        its original execution.
+
+    2. Set {{MyObject}}.[[{{started}}]] to `true`.
+
+    3. ...Do complex stuff...
+
+    4. [=Fire an event=] named `"currentlyinprogress"` at {{MyObject}}.
+
+    5. ...Finish up complex stuff...
+
+    6. Set {{MyObject}}.[[{{started}}]] to `false`.
+</div>
+
+Sometimes "guarding" as illustrated above cannot be done because of the complexity
+of the algorithm, the number of potential algorithm entry-points or the side-effects
+beyond the algorithm's control. In these cases another form of protecting the integrity
+of the algorithm's state is to "defer" the event to a subsequent task or microtask.
+Deferral ensures that any stack-based recursion is avoided (but does not eliminate
+potentially problematic loops, as they could now occur as unending follow-up tasks).
+
+Defering an event is often specified as "[=queue a task=] to [=fire an event=]...".
+
+By way of illustration, the deprecated
+[Mutation Events](https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Mutation_events),
+were specified to fire in the middle of node insertion and removal algorithms, which
+are triggered by a variety of common API calls like {{Node/appendChild()}}. The side-effects
+from recursivly running node mutation algorithms caused many years of security issues.
+A more robust approach is embodied in
+{{MutationObserver}}
+applies the principle of deferral of notification — in this case to the microtask queue
+following the algorithm.
+
+Weigh the trade-offs when considering deferral as an alternative to guarding as noted
+earlier. Guarding is generally the preferred approach. With guarding:
+
+* all state management can be rationalized within one "turn" of the algorithm (no
+    need to consider any state changes or API invocations between the time the algorithm
+    completes and when the event fires).
+* events can reveal internal state changes as early as possible when dispatched immediately
+    in the algorithm (resulting in minimal delay to user code that wants to act on
+    that state).
+
+Deferral as an option generally means *all* API entrypoints need guards, which is
+usually more work for the spec author. If there seems to be no need for guards, then
```

Comment:
I'm not sure what this sentence means. Is it getting at the fact that the state of the world may have changed between the state changing and the event firing, and thus you'll need more guards of some sort to deal with such state changes? It's not very clear...

Reply:
I spent some time thinking about what I meant--and it took a lot more words, but I think the result is better and more clear.